components/script/dom/globalscope.rs

```diff
@@ -285,6 +285,14 @@ struct FileListener {
     task_canceller: TaskCanceller,
 }
 
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct OffThreadCompilationData {
+    context: JSContext
+    canceller: TaskCanceller,
+    task_source: TimerTaskSource,
```

Comment:
To answer your other question, so here you should instead store the `NetworkingTaskSource`, which is available via a call to `GlobalScope.networking_task_source()`. 

Then later when you need to queue a task, you can do `OffThreadCompilationData.task_source.queue_with_canceller`, like is done for example at https://github.com/servo/servo/blob/d7d56454b07b70e3d9b2e3746ea1b2a1a1f05b7f/components/script/dom/globalscope.rs#L404

Reply:
I'd say you can define a `handle(&self, token: Token)` method on the struct, and then from the callback call `handle(token)`, and inside the method queue the task and move the token into the task closure, then inside the closure you can use the token for example to call `FinishOffThreadScript`.

Reply:
Once the `FinishOffThreadScript` has been called, we need to call `evaluate_script_on_global_with_result` with the obtained script, to call `JS_ExecuteScript`. So calling `CompileOffThread` can be called directly from `run_a_classic_script`. Is that correct?

Reply:
I think we need to:

1. Make `run_a_classic_script` take ownership of the `ScriptOrigin`.
2. Store the `ScriptOrigin` on `OffThreadCompilationData`, alongside the `Trusted<Global>`
3. Call the "compile off thread" API.

Then, from the task queued:

1. call `global.evaluate_script_on_global_with_result`.

`run_a_classic_script` doesn't return a result so that should be OK. 

You probably will have to change the signature of `evaluate_script_on_global_with_resul`, making the `code` argument an enum with  a `DomString` and a `JsScript` variant. 

Currently it's a `&str`, but it looks like it could in all current cases be a `DomString` instead.

Reply:
On second thoughts, I think we should do the off-thread compilation earlier, when the data is received from the network. 

So the entry point should I think be around here: https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L268

So I think you can:

Make `ScriptOrigin` have a new `Option<JsScript>` member named `compiled_script`.

1. Store both the `Trusted<HTMLScriptElement>` and the `ScriptOrigin`, and probably also `ExternalScriptKind` onto the `OffThreadCompilationData`.
2. Call the off-thread compilation API
3. From the callback, queue a task that will:
    a. call `FinishOffThreadScript`
    b. Set `compiled_script` to to the result of that call.
    c. continue at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L272

Then `evaluate_script_on_global_with_result` could be changed to take an `Option<JsScript>` as a new `compiled_script` argument, or maybe make the `code` into an enum, and inside you can either call `Evaluate2` or `JS_ExecuteScript`, based on what is available.

Reply:
Just a question, can these scripts be compiled in any context? As GlobalScope is not yet defined and would not be available.

Reply:
> Just a question, can these scripts be compiled in any context?

Yes I think so.

> Also how do we ensure in this that offthreadcompilation is used only when called by run_a_classic_script, as I believe this is done for all scripts?

I think the code at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L268 is only used for classic scripts, if not we can find another way to identify the script and either do the off-thread compilation or simply continue normally.

Reply:
Okay. So I think I will need to create a new task_canceller for this. Is it correct?

Reply:
So I'm almost certain that the code at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L269

is in fact only called for classic scripts, because of the name `ClassicContext` and the spec links, and I also think that is the right point to setup the `OffThreadCompilationData` and then call api to do the off-thread compilation. 

You can get the global at that point from `self.elem.global()`(since `elem` is `HTMLScriptElement`), and then you can get the JS context by doing `global.cx()`. 

Then you also need to store a `Trusted<HTMLScriptElement>` on the `OffThreadCompilationData`, meaning you don't need a `Trusted<GlobalScope>` because you can get the global from the element, and then later in the callback you queue the task that will then continue the algorithm at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L270

(where again you can get `elem` from the `Trusted<HTMLScriptElement>` you will have available by calling `root()` on it, actually the [`ClassicContext`](https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L186) used there to handle the network responses is very similar to the `OffThreadCompilationData` you will have to define. 

By the way, let's rename it `OffThreadCompilationContext`?

let me know if you have other questions.

Reply:
> o I think I will need to create a new task_canceller for this. Is it correct?

Yes you can call `global.task_canceller()` to get one.

Reply:
I tried getting the globalscope object using `self.elem.global()` but wasn't able to. `HTMLScriptElement` doesn't seem to have it. Can you please confirm?

Reply:
I think you first need to get the root out of the trusted, doing `let elem = self.elem.root(); `, then `elem.global()` should work

Reply:
I had some questions regarding OffThreadCompilation methods we are using.
1. These methods are written in C, but we do an unsafe call to them using a wrapper defined in rust?
2. For JSScript, I get the error that malloc_size_of is not implemented. Upon reading the documentation, I believe it is similar to a string in rust for the time being. So implementing that would mean something to similar done for string?
3. Another error during task queuing is OffThreadToken cannot be passed between threads safely. I think a simple clone should do the job, so that function needs to be implemented for it. Is that correct?

Reply:
1. Yes, I think that's a correct description.
2. You can ignore it, using an annoation like in https://github.com/servo/servo/blob/948dd7379618e85269227a3e37c76a091c85a966/components/script/dom/globalscope.rs#L250
3. I'm not sure about this one. I would say define a `struct OffthreadToken;`, and then in the callback cast the *mut OffThreadToken to it? Then cast it back when you call `FinishOffThreadScript`? @jdm do you have an idea how to represent `OffThreadToken* token` in Rust? It appears to be just an empty class defined at https://github.com/servo/mozjs/blob/aabcc9ba889b2755f1e4e83f28323a60415a790f/mozjs/js/src/vm/HelperThreads.h#L37